### ISSUE TITLE
Embedded Form upload not working due to output file permissions

### ISSUE THREAD

**Issue by pcbouman-eur** [autolab/Autolab#889]
When my colleague was experimenting with the use of embedded forms, he got the following error:

```
 Permission denied @ rb_sysopen - out.txt
  app/controllers/assessment/handin.rb:27:in `initialize'
```
Some minor experimentation on my side suggests that the "out.txt" file to which the form contents are written is located in the working directory of the Ruby process, as no path is specified. This location may not have write permissions in every setup, including our Docker based setup. Replacing line 27, which currently reads
```
      out_file = File.new("out.txt", "w+")
```
with the following trivial modification
```
      out_file = File.new("/tmp/out.txt", "w+")
```
solves the issue for us. Probably it would be even better to just rely on the "[create a temporary file for me](https://ruby-doc.org/stdlib-1.9.3/libdoc/tempfile/rdoc/Tempfile.html)" functionality of Ruby. Although I don't have experience with Ruby, and do not know how dependencies are organised in Rails, assuming that a Tempfile object can be used instead of a regular file object, something in the following lines could be a fix.
```
require 'tempfile'  # or some other way of doing this in Rails
# replace line 27 with the following
      out_file = Tempfile.new("out.txt", "w+")
```

**Comment by zyx-billy** [comment]
Yes, using a file was really a very ugly hack, as it may introduce race conditions (among other problems). I agree that using a ruby tempfile would be a good temporary fix to make it working, but in the long run we should set up a way to pass the contents directly to saveHandin (which actually saves the contents to a permanent file) without the need to create an intermediate file.

**Comment by skoch9** [comment]
The current version is dangerous to use (as it will definitely introduce problems when students use the platform in parallel) and should be fixed (or a warning should be added to the embedded forms documentation).

**Comment by pcbouman-eur** [comment]
The fix I have suggested at least fixes this race issue (the [ruby documentation](https://ruby-doc.org/stdlib-1.9.3/libdoc/tempfile/rdoc/Tempfile.html#class-Tempfile-label-Minor+notes) explicitly states that Tempfile.new is thread-safe and inter-process-safe) and should get rid of the fact that the embedded forms have been dangerous for at least a year now. As I don't have a proper Ruby dev setup, and am not really a Ruby developer any way I would prefer it if somebody else could turn this two-liner into a pull request. The restructuring of the saveHandin seems more like a feature/enhancement than a bugfix, imho.

**Comment by victorhuangwq** [comment]
Issue seems to be able to be addressed by this PR https://github.com/autolab/Autolab/pull/1140

**Comment by damianhxy** [comment]
Issue was resolved by #1441
